src/gui/spacemigration.cpp

```diff
@@ -0,0 +1,108 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+#include "spacemigration.h"
+
+#include "common/version.h"
+#include "gui/accountmanager.h"
+#include "gui/folderman.h"
+#include "libsync/account.h"
+#include "libsync/graphapi/drives.h"
+
+#include <QJsonArray>
+
+Q_LOGGING_CATEGORY(lcMigration, "gui.migration.spaces", QtInfoMsg)
+
+using namespace OCC;
+
+SpaceMigration::SpaceMigration(const AccountStatePtr &account, const QString &path, QObject *parent)
+    : QObject(parent)
+    , _accountState(account)
+    , _path(path)
+{
+}
+
+void SpaceMigration::start()
+{
+    FolderMan::instance()->setSyncEnabled(false);
+
+    QJsonArray folders;
+    for (auto *f : FolderMan::instance()->folders()) {
+        // same account and uses the legacy dav url
+        // already migrated folders are ignored
+        if (f->accountState()->account() == _accountState->account() && f->webDavUrl() == _accountState->account()->davUrl()) {
+            folders.append(f->remotePath());
+            _migrationFolders.append(f);
+        }
+    }
+    const QJsonObject obj {
+        { QStringLiteral("version"), Version::version().toString() },
+        { QStringLiteral("user"), _accountState->account()->davUser() },
+        { QStringLiteral("remotefolder"), folders }
+    };
+    auto job = new JsonJob(_accountState->account(), _accountState->account()->url(), _path, QByteArrayLiteral("POST"), obj);
+    job->setParent(this);
+    connect(job, &JsonJob::finishedSignal, this, [job, this] {
+        const int status = job->reply()->attribute(QNetworkRequest::HttpStatusCodeAttribute).toInt();
+        if (status == 200 && job->parseError().error == QJsonParseError::NoError) {
+            migrate(job->data().value(QStringLiteral("folders")).toObject());
+        } else {
+            qCInfo(lcMigration) << status << job->data();
+            if (job->parseError().error != QJsonParseError::NoError) {
+                qCInfo(lcMigration) << job->parseError().errorString();
+            }
+            Q_EMIT finised();
+        }
+    });
+    job->start();
+}
+
+void SpaceMigration::migrate(const QJsonObject &folders)
+{
+    auto drivesJob = new GraphApi::Drives(_accountState->account(), this);
+    connect(drivesJob, &GraphApi::Drives::finishedSignal, [drivesJob, folders, this] {
+        const auto drives = drivesJob->drives();
+        for (auto &folder : _migrationFolders) {
+            if (folder) {
+                const auto obj = folders.value(folder->remotePath()).toObject();
+                const QString error = obj.value(QStringLiteral("error")).toString();
+                if (error.isEmpty()) {
+                    const QString newPath = obj.value(QStringLiteral("path")).toString();
+                    const QString space_id = obj.value(QStringLiteral("space_id")).toString();
+                    auto it = std::find_if(drives.cbegin(), drives.cend(), [space_id](auto it) {
```

Comment:
I suppose the result will be `const` anyway? I guess this variable should be made `const` explicitly.